[BPF] Do atomic_fetch_*() pattern matching with memory ordering

yonghong-song opened the pull request:
Three commits in this pull request:
   commit 1: implement pattern matching for memory ordering seq_cst, acq_rel, release, acquire and monotonic. Specially, for monotonic memory ordering (relaxed memory model), if no return value is used, locked insn is used.
   commit 2: add support to handle dwarf atomic modifier in BTF generation. Actually atomic modifier is ignored in BTF.
   commit 3: add tests for new atomic ordering support and BTF support with _Atomic type.
I removed RFC tag as now patch sets are in reasonable states.

For atomic fetch_and_*() operations, do pattern matching with memory ordering
seq_cst, acq_rel, release, acquire and monotonic (relaxed). For fetch_and_*()
operations with seq_cst/acq_rel/release/acquire ordering, atomic_fetch_*()
instructions are generated. For monotonic ordering, locked insns are generated
if return value is not used. Otherwise, atomic_fetch_*() insns are used.
The main motivation is to resolve the kernel issue [1].
   
The following are memory ordering are supported:
  seq_cst, acq_rel, release, acquire, relaxed
Current gcc style __sync_fetch_and_*() operations are all seq_cst.

To use explicit memory ordering, the _Atomic type is needed.  The following is
an example:

```
$ cat test.c
\#include <stdatomic.h>
void f1(_Atomic int *i) {
   (void)__c11_atomic_fetch_and(i, 10, memory_order_relaxed);
}
void f2(_Atomic int *i) {
   (void)__c11_atomic_fetch_and(i, 10, memory_order_acquire);
}
void f3(_Atomic int *i) {
   (void)__c11_atomic_fetch_and(i, 10, memory_order_seq_cst);
}
$ cat run.sh
clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -c test.c -o test.o && llvm-objdum
p -d test.o
$ ./run.sh
       
test.o: file format elf64-bpf
       
Disassembly of section .text:

0000000000000000 <f1>:
       0:       b4 02 00 00 0a 00 00 00 w2 = 0xa
       1:       c3 21 00 00 50 00 00 00 lock *(u32 *)(r1 + 0x0) &= w2
       2:       95 00 00 00 00 00 00 00 exit
       
0000000000000018 <f2>:
       3:       b4 02 00 00 0a 00 00 00 w2 = 0xa
       4:       c3 21 00 00 51 00 00 00 w2 = atomic_fetch_and((u32 *)(r1 + 0x0), w2)
       5:       95 00 00 00 00 00 00 00 exit
       
0000000000000030 <f3>:
       6:       b4 02 00 00 0a 00 00 00 w2 = 0xa
       7:       c3 21 00 00 51 00 00 00 w2 = atomic_fetch_and((u32 *)(r1 + 0x0), w2)
       8:       95 00 00 00 00 00 00 00 exit
```    

The following is another example where return value is used:

```
$ cat test1.c
\#include <stdatomic.h>
int f1(_Atomic int *i) {
   return __c11_atomic_fetch_and(i, 10, memory_order_relaxed);
}  
int f2(_Atomic int *i) {
   return __c11_atomic_fetch_and(i, 10, memory_order_acquire);
}  
int f3(_Atomic int *i) {
   return __c11_atomic_fetch_and(i, 10, memory_order_seq_cst);
}  
$ cat run.sh
clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -c test1.c -o test1.o && llvm-objdump -d test1.o
$ ./run.sh

test.o: file format elf64-bpf

Disassembly of section .text:

0000000000000000 <f1>:
       0:       b4 00 00 00 0a 00 00 00 w0 = 0xa
       1:       c3 01 00 00 51 00 00 00 w0 = atomic_fetch_and((u32 *)(r1 + 0x0), w0)
       2:       95 00 00 00 00 00 00 00 exit
       
0000000000000018 <f2>:
       3:       b4 00 00 00 0a 00 00 00 w0 = 0xa
       4:       c3 01 00 00 51 00 00 00 w0 = atomic_fetch_and((u32 *)(r1 + 0x0), w0)
       5:       95 00 00 00 00 00 00 00 exit
       
0000000000000030 <f3>:
       6:       b4 00 00 00 0a 00 00 00 w0 = 0xa
       7:       c3 01 00 00 51 00 00 00 w0 = atomic_fetch_and((u32 *)(r1 + 0x0), w0)
       8:       95 00 00 00 00 00 00 00 exit
```    

You can see that for relaxed memory ordering, if return value is used, atomic_fetch_and()
insn is used. Otherwise, if return value is not used, locked insn is used.

Here is another example with global _Atomic variable:

```
$ cat test3.c
\#include <stdatomic.h>

_Atomic int i;

void f1(void) {
   (void)__c11_atomic_fetch_and(&i, 10, memory_order_relaxed);
}
void f2(void) {
   (void)__c11_atomic_fetch_and(&i, 10, memory_order_seq_cst);
}
$ cat run.sh
clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -c test3.c -o test3.o && llvm-objdump -d test3.o
$ ./run.sh

test3.o:        file format elf64-bpf

Disassembly of section .text:

0000000000000000 <f1>:
       0:       b4 01 00 00 0a 00 00 00 w1 = 0xa
       1:       18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r2 = 0x0 ll
       3:       c3 12 00 00 50 00 00 00 lock *(u32 *)(r2 + 0x0) &= w1
       4:       95 00 00 00 00 00 00 00 exit
       
0000000000000028 <f2>:
       5:       b4 01 00 00 0a 00 00 00 w1 = 0xa
       6:       18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r2 = 0x0 ll
       8:       c3 12 00 00 51 00 00 00 w1 = atomic_fetch_and((u32 *)(r2 + 0x0), w1)
       9:       95 00 00 00 00 00 00 00 exit
```    

Note that in the above compilations, '-g' is not used. The reason is due to the following IR
related to _Atomic type:
```
$clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -g -S -emit-llvm test3.c
```
The related debug info for test3.c:
```
!0 = !DIGlobalVariableExpression(var: !1, expr: !DIExpression())
!1 = distinct !DIGlobalVariable(name: "i", scope: !2, file: !3, line: 3, type: !16, isLocal: false, isDefinition: true)
...
!16 = !DIDerivedType(tag: DW_TAG_atomic_type, baseType: !17)
!17 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
```

If compiling test.c, the related debug info:
```
...
!19 = distinct !DISubprogram(name: "f1", scope: !1, file: !1, line: 3, type: !20, scopeLine: 3, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, retainedNodes: !25)
!20 = !DISubroutineType(types: !21)
!21 = !{null, !22}
!22 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: !23, size: 64)
!23 = !DIDerivedType(tag: DW_TAG_atomic_type, baseType: !24)
!24 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
!25 = !{!26}
!26 = !DILocalVariable(name: "i", arg: 1, scope: !19, file: !1, line: 3, type: !22)
```

All the above suggests _Atomic behaves like a modifier (e.g. const, restrict, volatile).
This seems true based on doc [1].

Without proper handling DW_TAG_atomic_type, llvm BTF generation will be incorrect since
the current implementation assumes no existence of DW_TAG_atomic_type. So we have
two choices here:
  (1). llvm bpf backend processes DW_TAG_atomic_type but ignores it in BTF encoding.
  (2). Add another type, e.g., BTF_KIND_ATOMIC to BTF. BTF_KIND_ATOMIC behaves as a
       modifier like const/volatile/restrict.

For choice (1), llvm bpf backend should skip dwarf::DW_TAG_atomic_type during
BTF generation whenever necessary.

For choice (2), BTF_KIND_ATOMIC will be added to BTF so llvm backend and kernel
needs to handle that properly. The main advantage of it probably is to maintain
this atomic type so it is also available to skeleton. But I think for skeleton
a raw type might be good enough unless user space intends to do some atomic
operation with that, which is a unlikely case.
    
So I choose choice (1) in this RFC implementation. See the commit message of the second commit for details.

 [1] https://lore.kernel.org/bpf/7b941f53-2a05-48ec-9032-8f106face3a3@linux.dev/
 [2] https://dwarfstd.org/issues/131112.1.html

Commit:
[BPF] Do atomic_fetch_*() pattern matching with memory ordering

For atomic fetch_and_*() operations, do pattern matching with memory ordering
seq_cst, acq_rel, release, acquire and monotonic (relaxed). For fetch_and_*()
operations with seq_cst/acq_rel/release/acquire ordering, atomic_fetch_*()
instructions are generated. For monotonic ordering, locked insns are generated
if return value is not used. Otherwise, atomic_fetch_*() insns are used.
The main motivation is to resolve the kernel issue [1].

The following are memory ordering are supported:
  seq_cst, acq_rel, release, acquire, relaxed
Current gcc style __sync_fetch_and_*() operations are all seq_cst.

To use explicit memory ordering, the _Atomic type is needed.  The following is
an example:

```
$ cat test.c
\#include <stdatomic.h>
void f1(_Atomic int *i) {
   (void)__c11_atomic_fetch_and(i, 10, memory_order_relaxed);
}
void f2(_Atomic int *i) {
   (void)__c11_atomic_fetch_and(i, 10, memory_order_acquire);
}
void f3(_Atomic int *i) {
   (void)__c11_atomic_fetch_and(i, 10, memory_order_seq_cst);
}
$ cat run.sh
clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -c test.c -o test.o && llvm-objdump -d test.o
$ ./run.sh

test.o: file format elf64-bpf

Disassembly of section .text:

0000000000000000 <f1>:
       0:       b4 02 00 00 0a 00 00 00 w2 = 0xa
       1:       c3 21 00 00 50 00 00 00 lock *(u32 *)(r1 + 0x0) &= w2
       2:       95 00 00 00 00 00 00 00 exit

0000000000000018 <f2>:
       3:       b4 02 00 00 0a 00 00 00 w2 = 0xa
       4:       c3 21 00 00 51 00 00 00 w2 = atomic_fetch_and((u32 *)(r1 + 0x0), w2)
       5:       95 00 00 00 00 00 00 00 exit

0000000000000030 <f3>:
       6:       b4 02 00 00 0a 00 00 00 w2 = 0xa
       7:       c3 21 00 00 51 00 00 00 w2 = atomic_fetch_and((u32 *)(r1 + 0x0), w2)
       8:       95 00 00 00 00 00 00 00 exit
```

The following is another example where return value is used:

```
$ cat test1.c
\#include <stdatomic.h>
int f1(_Atomic int *i) {
   return __c11_atomic_fetch_and(i, 10, memory_order_relaxed);
}
int f2(_Atomic int *i) {
   return __c11_atomic_fetch_and(i, 10, memory_order_acquire);
}
int f3(_Atomic int *i) {
   return __c11_atomic_fetch_and(i, 10, memory_order_seq_cst);
}
$ cat run.sh
clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -c test1.c -o test1.o && llvm-objdump -d test1.o
$ ./run.sh

test.o: file format elf64-bpf

Disassembly of section .text:

0000000000000000 <f1>:
       0:       b4 00 00 00 0a 00 00 00 w0 = 0xa
       1:       c3 01 00 00 51 00 00 00 w0 = atomic_fetch_and((u32 *)(r1 + 0x0), w0)
       2:       95 00 00 00 00 00 00 00 exit

0000000000000018 <f2>:
       3:       b4 00 00 00 0a 00 00 00 w0 = 0xa
       4:       c3 01 00 00 51 00 00 00 w0 = atomic_fetch_and((u32 *)(r1 + 0x0), w0)
       5:       95 00 00 00 00 00 00 00 exit

0000000000000030 <f3>:
       6:       b4 00 00 00 0a 00 00 00 w0 = 0xa
       7:       c3 01 00 00 51 00 00 00 w0 = atomic_fetch_and((u32 *)(r1 + 0x0), w0)
       8:       95 00 00 00 00 00 00 00 exit
```

You can see that for relaxed memory ordering, if return value is used, atomic_fetch_and()
insn is used. Otherwise, if return value is not used, locked insn is used.

Here is another example with global _Atomic variable:

```
$ cat test3.c
\#include <stdatomic.h>

_Atomic int i;

void f1(void) {
   (void)__c11_atomic_fetch_and(&i, 10, memory_order_relaxed);
}
void f2(void) {
   (void)__c11_atomic_fetch_and(&i, 10, memory_order_seq_cst);
}
$ cat run.sh
clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -c test3.c -o test3.o && llvm-objdump -d test3.o
$ ./run.sh

test3.o:        file format elf64-bpf

Disassembly of section .text:

0000000000000000 <f1>:
       0:       b4 01 00 00 0a 00 00 00 w1 = 0xa
       1:       18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r2 = 0x0 ll
       3:       c3 12 00 00 50 00 00 00 lock *(u32 *)(r2 + 0x0) &= w1
       4:       95 00 00 00 00 00 00 00 exit

0000000000000028 <f2>:
       5:       b4 01 00 00 0a 00 00 00 w1 = 0xa
       6:       18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r2 = 0x0 ll
       8:       c3 12 00 00 51 00 00 00 w1 = atomic_fetch_and((u32 *)(r2 + 0x0), w1)
       9:       95 00 00 00 00 00 00 00 exit
```

Note that in the above compilations, '-g' is not used. The reason is due to the following IR
related to _Atomic type:
```
$clang  -I/home/yhs/work/bpf-next/tools/testing/selftests/bpf -O2 --target=bpf -g -S -emit-llvm test3.c
```
The related debug info for test3.c:
```
!0 = !DIGlobalVariableExpression(var: !1, expr: !DIExpression())
!1 = distinct !DIGlobalVariable(name: "i", scope: !2, file: !3, line: 3, type: !16, isLocal: false, isDefinition: true)
...
!16 = !DIDerivedType(tag: DW_TAG_atomic_type, baseType: !17)
!17 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
```

If compiling test.c, the related debug info:
```
...
!19 = distinct !DISubprogram(name: "f1", scope: !1, file: !1, line: 3, type: !20, scopeLine: 3, flags: DIFlagPrototyped | DIFlagAllCallsDescribed, spFlags: DISPFlagDefinition | DISPFlagOptimized, unit: !0, retainedNodes: !25)
!20 = !DISubroutineType(types: !21)
!21 = !{null, !22}
!22 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: !23, size: 64)
!23 = !DIDerivedType(tag: DW_TAG_atomic_type, baseType: !24)
!24 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
!25 = !{!26}
!26 = !DILocalVariable(name: "i", arg: 1, scope: !19, file: !1, line: 3, type: !22)
```

All the above suggests _Atomic behaves like a modifier (e.g. const, restrict, volatile).
This seems true based on doc [1].

Without proper handling DW_TAG_atomic_type, llvm BTF generation will be incorrect since
the current implementation assumes no existence of DW_TAG_atomic_type. So we have
two choices here:
  (1). llvm bpf backend processes DW_TAG_atomic_type but ignores it in BTF encoding.
  (2). Add another type, e.g., BTF_KIND_ATOMIC to BTF. BTF_KIND_ATOMIC behaves as a
       modifier like const/volatile/restrict.

For choice (1), llvm bpf backend should skip dwarf::DW_TAG_atomic_type during
BTF generation whenever necessary.

For choice (2), BTF_KIND_ATOMIC will be added to BTF so llvm backend and kernel
needs to handle that properly. The main advantage of it probably is to maintain
this atomic type so it is also available to skeleton. But I think for skeleton
a raw type might be good enough unless user space intends to do some atomic
operation with that, which is a unlikely case.

So I choose choice (1) in this implementation.

 [1] https://lore.kernel.org/bpf/7b941f53-2a05-48ec-9032-8f106face3a3@linux.dev/
 [2] https://dwarfstd.org/issues/131112.1.html

yonghong-song commented:
cc @anakryiko @jemarch

eddyz87 commented:
So, basically we want to generate `lock *(u64 *)(rX + ...) += rY` for `__c11_atomic_fetch_and(..., memory_order_relaxed)` and `rX = atomic_fetch_add(...)` for everything else.
`memory_order_relaxed` corresponds to [monotonic](https://llvm.org/docs/Atomics.html#monotonic) LLVM IR ordering.
And basing on the referred mailing list discussion, this is done to allow ARM jit to generate [`LDADD`](https://developer.arm.com/documentation/ddi0602/2024-06/Base-Instructions/LDADD--LDADDA--LDADDAL--LDADDL--Atomic-add-on-word-or-doubleword-in-memory-) instruction (w/o `L` or `A` or `AL` suffixes) for this C interface (which it currently does for `lock ...` instructions). 

I was unable to figure out from the ARM documentation whether `LDADD` is [monotonic](https://llvm.org/docs/Atomics.html#monotonic) or [unordered](https://llvm.org/docs/Atomics.html#unordered). However, the test below shows that at-least we are on the same page with LLVM ARM backend:

```c
$ cat test2.c
void f1(_Atomic long *i) { __c11_atomic_fetch_add(i, 10, __ATOMIC_RELAXED); }
void f2(_Atomic long *i) { __c11_atomic_fetch_add(i, 10, __ATOMIC_CONSUME); }
void f3(_Atomic long *i) { __c11_atomic_fetch_add(i, 10, __ATOMIC_ACQUIRE); }
void f4(_Atomic long *i) { __c11_atomic_fetch_add(i, 10, __ATOMIC_RELEASE); }
void f5(_Atomic long *i) { __c11_atomic_fetch_add(i, 10, __ATOMIC_ACQ_REL); }
void f6(_Atomic long *i) { __c11_atomic_fetch_add(i, 10, __ATOMIC_SEQ_CST); }
```
```bash
$ clang --target=aarch64 -march=armv8.1-a -O2 test2.c -c -o - | llvm-objdump -Sdr -
```
```asm
<stdin>:	file format elf64-littleaarch64

Disassembly of section .text:

0000000000000000 <f1>:
       0: 52800148     	mov	w8, #0xa                // =10
       4: f8280008     	ldadd	x8, x8, [x0]
       8: d65f03c0     	ret

000000000000000c <f2>:
       c: 52800148     	mov	w8, #0xa                // =10
      10: f8a80008     	ldadda	x8, x8, [x0]
      14: d65f03c0     	ret
0000000000000018 <f3>:
      18: 52800148     	mov	w8, #0xa                // =10
      1c: f8a80008     	ldadda	x8, x8, [x0]
      20: d65f03c0     	ret
0000000000000024 <f4>:
      24: 52800148     	mov	w8, #0xa                // =10
      28: f8680008     	ldaddl	x8, x8, [x0]
      2c: d65f03c0     	ret
0000000000000030 <f5>:
      30: 52800148     	mov	w8, #0xa                // =10
      34: f8e80008     	ldaddal	x8, x8, [x0]
      38: d65f03c0     	ret
000000000000003c <f6>:
      3c: 52800148     	mov	w8, #0xa                // =10
      40: f8e80008     	ldaddal	x8, x8, [x0]
      44: d65f03c0     	ret
```

So, I think we are good.

Question: should BPF backend report and error if `__ATOMIC_{CONSUME,ACQUIRE,RELEASE,ACQ_REL}` is used?
LLVM documentation [allows this](https://llvm.org/docs/Atomics.html#unordered).

eddyz87 commented:
Also note:

```c
$ cat test2.c
long f1(_Atomic long *i) { return __c11_atomic_fetch_add(i, 10, __ATOMIC_RELAXED); }
```
```c
$ clang --target=bpf -mcpu=v3 -O2 test2.c -c -o - | llvm-objdump -Sdr -
test2.c:1:6: error: Invalid usage of the XADD return value
    1 | long f1(_Atomic long *i) { return __c11_atomic_fetch_add(i, 10, __ATOMIC_RELAXED); }
      |      ^
1 error generated.

<stdin>:	file format elf64-bpf

Disassembly of section .text:

0000000000000000 <f1>:
       0:	b7 00 00 00 0a 00 00 00	r0 = 0xa
       1:	db 01 00 00 00 00 00 00	lock *(u64 *)(r1 + 0x0) += r0
       2:	95 00 00 00 00 00 00 00	exit
```

Which seems incorrect.

yonghong-song commented:
> Question: should BPF backend report and error if `__ATOMIC_{CONSUME,ACQUIRE,RELEASE,ACQ_REL}` is used?
> LLVM documentation [allows this](https://llvm.org/docs/Atomics.html#unordered).
> Edit: currently `clang` crashes with a backtrace if one of these is used, probably better to report an error in a more user friendly way.

Agree that Issuing an error message is more user friendly. Also __ATOMIC_CONSUME is an variant of __ATOMIC_ACQUIRE but looks like they may have subtle difference. In https://gcc.gnu.org/onlinedocs/gcc-4.9.2/gcc/_005f_005fatomic-Builtins.html
```
__ATOMIC_CONSUME
    Data dependency only for both barrier and synchronization with another thread.
__ATOMIC_ACQUIRE
    Barrier to hoisting of code and synchronizes with release (or stronger) semantic stores from another thread. 
```
Linux kernel does not support __ATOMIC_CONSUME. So BPF backend may not support it anytime soon.

yonghong-song commented:
> Also note:
> 
> ```c
> $ cat test2.c
> long f1(_Atomic long *i) { return __c11_atomic_fetch_add(i, 10, __ATOMIC_RELAXED); }
> ```
> 
> ```c
> $ clang --target=bpf -mcpu=v3 -O2 test2.c -c -o - | llvm-objdump -Sdr -
> test2.c:1:6: error: Invalid usage of the XADD return value
>     1 | long f1(_Atomic long *i) { return __c11_atomic_fetch_add(i, 10, __ATOMIC_RELAXED); }
>       |      ^
> 1 error generated.
> 
> <stdin>:	file format elf64-bpf
> 
> Disassembly of section .text:
> 
> 0000000000000000 <f1>:
>        0:	b7 00 00 00 0a 00 00 00	r0 = 0xa
>        1:	db 01 00 00 00 00 00 00	lock *(u64 *)(r1 + 0x0) += r0
>        2:	95 00 00 00 00 00 00 00	exit
> ```
> 
> Which seems incorrect.

You are right. Yes, if the func return value is used, we should just use atomic_fetch_*() insns instead of locked insns.

anakryiko commented:
> Question: should BPF backend report and error if `__ATOMIC_{CONSUME,ACQUIRE,RELEASE,ACQ_REL}` is used?

Why not use the strongest ordering for all those "intermediate" orderings instead of erroring out? Just use `__ATOMIC_SEQ_CST`? If, in the future, BPF has weaker orderings, then we can always downgrade to that (probably with cpu version bump).

eddyz87 commented:
> > Question: should BPF backend report and error if `__ATOMIC_{CONSUME,ACQUIRE,RELEASE,ACQ_REL}` is used?
> 
> Why not use the strongest ordering for all those "intermediate" orderings instead of erroring out? Just use `__ATOMIC_SEQ_CST`? If, in the future, BPF has weaker orderings, then we can always downgrade to that (probably with cpu version bump).

I assume that there would be some reasoning behind users asking for a specific memory order (e.g. hope for better performance), and thus it's better to let user know that which orderings are supported.

But falling back to `__ATOMIC_SEQ_CST` is an option, yes.

anakryiko commented:
Not all orderings are meaningful on all architectures, but compilers don't error out on that, no? Think about writers of some common helpers dealing with atomic primitives. Why should they worry about which specific mode is supported (especially that that might change over time depending on *kernel version*)? They request the *minimum* level of ordering, but they would equally well work with more strict ordering, if the underlying architecture (verifier/JIT in this case) doesn't support some of them on some older versions.

eddyz87 commented:
> More broadly speaking (non-BPF specific), if I know I'm ok with relaxed ordering, but some architecture only supports SEQ_CST, I'm fine with that as a writer of atomic code. It's much bigger PITA to work around all these version/arch specific issues if they become compile-time or runtime errors, IMO.

Fair enough.
@yonghong-song , what do you think?

yonghong-song commented:
@anakryiko @eddyz87 Yes, we can implement other memory ordering (acquire, release, act_rel) to be the same as seq_cst initially. Later on, we can invent new instructions in llvm to actually support acquire, release, act_rel properly. It is very likely that when we are able to land llvm/kernel change, we may already figure out the new insn format and we will just implement them at that time.

yonghong-song commented:
Just made the change to support seq_cst/acq_rel/acquire/release/monotonic memory ordering. Also fixed the issue where locked insn is used when actual source expects a return value. I will continue to work on llvm BPF backend to process DW_TAG_atomic_type in the next step.

Commit:
[BPF] Handle DW_TAG_atomic_type properly

Make change in BTFDebug.cpp to handle DW_TAG_atomic_type properly.
Otherwise, a type like
   _Atomic int i; // global
the dwarf type chain atomic->int
Since DW_TAG_atomic_type is not processed BTF generation will stop
at atomic modifier and BTF will encode 'i' as void type.

Similar for type like
  volatile _Atomic int *p;
the dwarf type chain ptr->volatile->atomic->int
Since atomic type is not processed and BTF generation will stop at
atomic type, the eventual BTF type will be
  ptr->volatile->void
which is incorrect.

This patch fixed the following cases including the above two patterns
by skipping DW_TAG_atomic_type:
  - global variable with _Atomic type.
  - function parameter and return type with _Atomic type.
  - struct member with _Atomic type.
  - ptr,const,volatile,restrict pointing to a _Atomic type.
  - btf_type_tag where ptr pointing to _Atomic type and btf_type_tag.

With changed llvm, in kernel selftest arena_atomics.c ([1]), the new bpf
code looks like

```
_Atomic __u64 __arena_global and64_value = (0x110ull << 32);
_Atomic __u32 __arena_global and32_value = 0x110;

SEC("raw_tp/sys_enter")
int and(const void *ctx)
{
	...
        __c11_atomic_fetch_and(&and64_value, 0x011ull << 32, memory_order_relaxed);
        __c11_atomic_fetch_and(&and32_value, 0x011, memory_order_relaxed);
	...

        return 0;
}
```
and compilation is successful.

The skel file arena_atomics.skel.h will be
```
struct arena_atomics__arena {
	...
	__u64 and64_value;
	__u32 and32_value;
	...
} *arena;
```

  [1] https://lore.kernel.org/r/20240909223431.1666305-1-yonghong.song@linux.dev

yonghong-song commented:
Add change in BTFDebug.cpp to ingore dwarf::DW_TAG_atomic_type. Add some info about bpf prog atomic type usage and what skeleton looks like (see commit message in commit 2).

eddyz87 commented on llvm/lib/Target/BPF/BPFInstrInfo.td:
```diff
@@ -864,26 +864,119 @@ class XFALU32<BPFWidthModifer SizeOp, BPFArithOp Opc, string OpcodeStr,
+}
+
+let Predicates = [BPFHasALU32] in {
+    def : Pat<(atomic_load_add_i32_monotonic ADDRri:$addr, GPR32:$val),
```
We guard add operation with `BPFHasALU32`, but not or/xor/and, why?

yonghong-song replied:
I guess indeed I need to guard thos or/xor/and 32-bit atomic operations with BPFHasALU32.

eddyz87 commented on llvm/lib/Target/BPF/BPFInstrInfo.td:
```diff
@@ -864,26 +864,119 @@ class XFALU32<BPFWidthModifer SizeOp, BPFArithOp Opc, string OpcodeStr,
 // atomic_load_sub can be represented as a neg followed
 // by an atomic_load_add.
-def : Pat<(atomic_load_sub_i32 ADDRri:$addr, GPR32:$val),
+// FIXME: the below can probably be simplified.
```
E.g. use a `foreach` as in this [patch](https://github.com/user-attachments/files/16938017/foreach.diff.txt), wdyt?
There is probably a way to use `multiclass` to factor out tuples `["i32", "W32", GPR32]` and `["i64", "D", GPR]`  but my tablegen skills were not sufficient enough :)

yonghong-song replied:
Good suggestion. I know it must be a way to simplify it and I just didn't do it yet. Will give a try later.

yonghong-song replied:
> E.g. use a `foreach` as in this [patch](https://github.com/user-attachments/files/16938017/foreach.diff.txt), wdyt? There is probably a way to use `multiclass` to factor out tuples `["i32", "W32", GPR32]` and `["i64", "D", GPR]` but my tablegen skills were not sufficient enough :)

Thanks Eduard! Your `foreach` approach looks good to me. I try to do more simplification but is not successful. So literally I just used your implementation in my code.

eddyz87 replied:
Understood, thank you. At times wrangling tablegen to do what one wants takes an unreasonable amount of time :(

eddyz87 commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -1444,8 +1454,15 @@ void BTFDebug::processGlobals(bool ProcessingMapDef) {
-        visitTypeEntry(DIGlobal->getType(), GVTypeId, false, false);
+      else {
+        const DIType *Ty = DIGlobal->getType();
+        if (Ty) {
```
Silly question, when can `Ty` be null?

eddyz87 replied:
Understood, thank you

eddyz87 commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -91,6 +91,12 @@ void BTFTypeDerived::completeType(BTFDebug &BDebug) {
   const DIType *ResolvedType = DTy->getBaseType();
+  if (ResolvedType) {
+    const auto *DerivedTy = dyn_cast<DIDerivedType>(ResolvedType);
+    if (DerivedTy && DerivedTy->getTag() == dwarf::DW_TAG_atomic_type)
```
Nit: since this is done in two places, maybe extract it as a utility?

yonghong-song replied:
Sure. Will try to clean up the code later.

eddyz87 commented on llvm/lib/Target/BPF/BPFMIChecking.cpp:
```diff
@@ -152,22 +152,91 @@ static bool hasLiveDefs(const MachineInstr &MI, const TargetRegisterInfo *TRI) {
   for (MachineBasicBlock &MBB : *MF) {
     for (MachineInstr &MI : MBB) {
-      if (MI.getOpcode() != BPF::XADDW && MI.getOpcode() != BPF::XADDD)
+      if (ToErase) {
```
This is needed to allow iterator move to the next instruction, right?
A bit tricky, maybe add a comment here?
Also, won't delete the last instruction in the basic block, would this be a problem if instruction is used in a later basic block?

yonghong-song replied:
Will add a comment.
For
> Also, won't delete the last instruction in the basic block, would this be a problem if instruction is used in a later basic block?
I think we should be okay since register definition is preserved. But let me double check.

eddyz87 reviewed (commented):
I agree with these changes, could you please also add tests?

yonghong-song commented:
For
> I agree with these changes, could you please also add tests?

Sure. Will do.

Commit:
[BPF] Add functionality/btf selftests for memory ordering cases

The following test cases are added:
  - all memory ordering and its asm codes with -mcpu=v3
  - all memory ordering and its asm codes with -mcpu=v1
    Note that __c11_atomic_fetch_{sub,and,or,xor} for 32bit won't
    work for -mcpu=v1. Also at -mcpu=v1, no return value
    allowed for 64bit __sync_fetch_and_add.
  - at -mcpu=v1, __c11_atomic_fetch_sub() for 64bit with relaxed
    memory ordering, the xaddd insn will be used so return
    value is not supported. Otherwise, it will work fine
    if return value is not used. This aligns to
    __c11_atomic_fetch_add() for 64bit with relaxed memory
    ordering at -mcpu=v1.
  - BTF test with _Atomic types in different cases.

yonghong-song commented:
This pull request has been used for the merged kernel patch:
   https://lore.kernel.org/r/20240909223431.1666305-1-yonghong.song@linux.dev

eddyz87 commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -35,6 +35,15 @@ static const char *BTFKindStr[] = {
 #include "llvm/DebugInfo/BTF/BTF.def"
 };
 
+static const DIType *tryRemoveAtomicType(const DIType *Ty) {
```
Is it necessary to call this thing in `BTFTypeTypeTag::completeType`?

yonghong-song replied:
Yes. It is necessary. completeType intends to find the  *next* type id. For example, 'volatile atomic int', we want to find the target type id of 'volatile ...', we get the base type of 'volatile' and then through hash table to get the id.
But we want to skip atomic if it is there. That is why we need to call tryRemoveAtomicType.

eddyz87 replied:
I meant that it seems that `tryRemoveAtomicType` is missing from `BTFTypeTypeTag::completeType`, e.g. the following example drops type tag:

```
$ cat test5.c 
int _Atomic __attribute__((btf_type_tag("foo"))) *root;

$ clang -c --target=bpf -g test5.c -O2 -o test5.o && bpftool btf dump file test5.o
[1] PTR '(anon)' type_id=2
[2] INT 'int' size=4 bits_offset=0 nr_bits=32 encoding=SIGNED
[3] VAR 'root' type_id=1, linkage=global
[4] DATASEC '.bss' size=0 vlen=1
	type_id=3 offset=0 size=8 (VAR 'root')
```

As far as I understand, expected output is:

```
[1] TYPE_TAG 'foo' type_id=3
[2] PTR '(anon)' type_id=1
[3] INT 'int' size=4 bits_offset=0 nr_bits=32 encoding=SIGNED
[4] VAR 'root' type_id=2, linkage=global
[5] DATASEC '.bss' size=0 vlen=1
	type_id=4 offset=0 size=8 (VAR 'root')
```

yonghong-song replied:
Good point. I probably missed this case. Let me fix it.

eddyz87 commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -800,6 +809,10 @@ void BTFDebug::visitDerivedType(const DIDerivedType *DTy, uint32_t &TypeId,
+  if (Tag == dwarf::DW_TAG_atomic_type)
+    return visitTypeEntry(DTy->getBaseType(), TypeId, CheckPointer,
+                          SeenPointer);
+
```
Below there is a check:

```c++
if (CheckPointer && SeenPointer) {
    const DIType *Base = DTy->getBaseType();
if (Base) {
  ...
```

Is it necessary to skip atomic there?
E.g. to handle situation like below:

```c
struct foo {
  void * _Atomic a;
} root;

/*
$ clang -O2 -g -mcpu=v3 --target=bpf -c test4.c -o test4.o && bpftool btf dump file test4.o
[1] STRUCT 'foo' size=8 vlen=1
	'a' type_id=2 bits_offset=0
[2] PTR '(anon)' type_id=0
[3] VAR 'root' type_id=1, linkage=global
[4] DATASEC '.bss' size=0 vlen=1
	type_id=3 offset=0 size=8 (VAR 'root')
*/
```

eddyz87 commented on llvm/test/CodeGen/BPF/BTF/atomics.ll:
```diff
@@ -0,0 +1,186 @@
+  ret i32 %a1, !dbg !55
+}
+
+; CHECK:             .long   1                               # BTF_KIND_INT(id = 1)
```
Nit: such tests are extremely hard to follow. I suggest to reuse my old [python script](https://github.com/llvm/llvm-project/blob/14033e79d9b4988807145e52540a76889b477459/llvm/test/CodeGen/BPF/BTF/print_btf.py#L1) that hadn't made it into `main` yet.

yonghong-song replied:
Okay, let me take a look at this. Thanks for the reference.

eddyz87 replied:
Looks like linter does not like the way the script is formatted, sorry about that :(
I can update the script and attach here, if you'd like.

yonghong-song replied:
No problem. Let me fix it.

eddyz87 commented:
@yonghong-song, note: if all three commits remain a part of a single pull request, the are required to be squashed ([link](https://llvm.org/docs/GitHub.html)). If you want all three to be separate, each has to have it's separate pull request (one branch forked from another).

yonghong-song commented on llvm/lib/Target/BPF/BPFMIChecking.cpp:
```diff
@@ -152,22 +152,93 @@ static bool hasLiveDefs(const MachineInstr &MI, const TargetRegisterInfo *TRI) {
+        }
+      }
+    }
+  }
```
Good question. I should fail the compilation indeed.

eddyz87 replied:
In BPFISelLowering.cpp we have function `fail()`, which reports error message/location, prints a backtrace and does `exit(1)` (via default diagnostic handler). Which is not ideal either. Oh, well.

yonghong-song commented on llvm/lib/Target/BPF/BPFMIChecking.cpp:
```diff
@@ -152,22 +152,93 @@ static bool hasLiveDefs(const MachineInstr &MI, const TargetRegisterInfo *TRI) {
+    }
+  }
+
+  // Check return values of atomic_fetch_and_{add,and,or,xor}.
```
binop_no_use and binop_has_use are indeed much better. Let me give a try.

yonghong-song commented:
> @yonghong-song, note: if all three commits remain a part of a single pull request, the are required to be squashed ([link](https://llvm.org/docs/GitHub.html)). If you want all three to be separate, each has to have it's separate pull request (one branch forked from another).

I will squash them into a single commit.

yonghong-song commented:
Upload a new version:
  - better pattern matching (differentiating return vs. no-return) to avoid late optimization hack (from @eddyz87)
  - use python script print_btf.py to dump BTF which is easier to understand (from @eddyz87)

eddyz87 commented on llvm/test/CodeGen/BPF/atomics_mem_order_v1.ll:
```diff
@@ -0,0 +1,385 @@
+}
+
+; Function Attrs: mustprogress nofree norecurse nounwind willreturn memory(argmem: readwrite)
+define dso_local void @test_fetch_add_64_noret(ptr nocapture noundef %i) local_unnamed_addr #0 {
```
The difference between `add` and `sub` handling in `@test_fetch_add_64_noret` and `@test_fetch_sub_64_noret` looks a bit surprising.

yonghong-song replied:
good question. I am aware of this difference. In llvm19 and earlier releases, we have
```
// atomic_load_sub can be represented as a neg followed
// by an atomic_load_add.
def : Pat<(atomic_load_sub_i32 ADDRri:$addr, GPR32:$val),
          (XFADDW32 ADDRri:$addr, (NEG_32 GPR32:$val))>;
def : Pat<(atomic_load_sub_i64 ADDRri:$addr, GPR:$val),
          (XFADDD ADDRri:$addr, (NEG_64 GPR:$val))>;
```

For atomic_load_sub_i64, we use XFADDD for v1 instead of XADDD. So we have decided to have atomic_load_sub_i64 similar to atomic_load_{and,or,xor}_i64 and in this patch I didn't make that change.

eddyz87 replied:
Oh, ok.

yonghong-song commented:
Uploaded new codes to address the missing case like
```
int _Atomic __attribute__((btf_type_tag("foo"))) *root;
```

uweigand commented:
Looks like the new test case is failing on SystemZ: https://lab.llvm.org/buildbot/#/builders/42/builds/1192
```
struct.error: unpack_from requires a buffer of at least 402653196 bytes for unpacking 12 bytes at offset 402653184 (actual buffer size is 479)
```
At first glance, this might be an endian problem (SystemZ is big-endian; 402653196 is 0x1800000c).

yonghong-song commented:
> Looks like the new test case is failing on SystemZ: https://lab.llvm.org/buildbot/#/builders/42/builds/1192
> 
> ```
> struct.error: unpack_from requires a buffer of at least 402653196 bytes for unpacking 12 bytes at offset 402653184 (actual buffer size is 479)
> ```
> 
> At first glance, this might be an endian problem (SystemZ is big-endian; 402653196 is 0x1800000c).

Thanks for reporting! The test explicitly specified little endian like below:
```
; RUN: llc -march=bpfel -mcpu=v3 -filetype=obj -o %t1 %s
; RUN: llvm-objcopy --dump-section='.BTF'=%t2 %t1
; RUN: %python %p/print_btf.py %t2 | FileCheck -check-prefixes=CHECK %s
```
Let me investigate what is the problem here.

eddyz87 commented:
@yonghong-song , @uweigand , the fix is available in #110332
